### DISCUSSION TITLE
Julia WebAssembly backend

### DISCUSSION THREAD

**Discussion by pankgeorg** [fonsp/Pluto.jl#572] [Pluto development]
Web Assembly exists for this exact use case: client downloads a bundle and runs the whole Julia system within a wasm process.

That will make sharing work much easier since only data and assets will need to be hosted somewhere (no external computing power needed).

What do you think about that? I can investigate further if it makes sense to you.

**Comment by fonsp** [comment]
Last I checked, Julia isn't ready yet, but feel free to experiment!

I believe that keno is the person driving wasm support, and he recently stated that he will work on it further and use Pluto as frontend: https://www.heise.de/news/Data-Science-Mozilla-stellt-webbasiertes-Tool-Iodide-ein-4918368.html

**Comment by dralletje** [comment]
https://keno.github.io/julia-wasm/website/repl.htm looks very cool though

**Comment by fonsp** [comment]
@pankgeorg Feel free to experiment and tell us what you found

**Comment by fonsp** [comment]
Running Pluto on WebAssembly would be a game changer - a huge leap in accessible and collaborative science! You could run & share notebooks entirely in the browser, without depending on a commercial service like google colab. To get there, we need to run Julia on wasm.

I talked with @keno, he is leading this development (https://github.com/Keno/julia-wasm) and he is excited to use Pluto as front-end for the first working demo. We could help him get there! Mainly:

### 1. Serious compiler business
keno told me that someone needs to dig into https://github.com/Keno/julia-wasm and get it running again with the up to date versions. Consider doing this in a nextjournal.com bash notebook so that others can reproduce.

### 2. Rewrite parts of Pluto to make it run in the browser -- EDIT: this is [done](https://pluto-wasm-backend.netlify.app/editor.html) ([code](https://github.com/fonsp/Pluto.jl/tree/wasm-backend-1))
A first step would be to run the frontend standalone, with a JS interpreter as backend for example. This would make it possible to run a feature-incomplete julia-wasm already, non-reactively (i.e. the current REPL demo wrapped in our UI). This would be a good demo to get more attention.

---

If anyone is interested, let me know! It would be super super cool to get this going. We also have a budget to pay you to work on this :) fons@plutojl.org if you are interested!

**Comment by fonsp** [comment]
I made a fork https://github.com/fonsp/julia-wasm with everything removed except a Julia runtime in the browser console

**Reply by VarLad** [reply]
Any progress on this...?

**Reply by fonsp** [reply]
Yes! I updated my previous comment -- https://pluto-wasm-backend.netlify.app/editor.html

**Reply by VarLad** [reply]
@fonsp 
I just opened it on my phone
1. It opened instantaneously
2. I ran the blocks and they do work
.....
Its magic for me at this point...
Thank you(and Keno) for your efforts🙂

**Comment by logankilpatrick** [comment]
Is this a stable replacement for https://julialang.org/learning/tryjulia/ and Replit? It would be really cool to get Pluto there instead.

**Reply by fonsp** [reply]
Unfortunately not yet! WASM would be amazing for this, I think good WASM support would let Julia reach so many new people, but it is not stable enough yet, and not getting a lot of attention from the Julia devs.

Anyway, I can make a binder-based Pluto link instead if you like? It would be about 20-30 seconds launch time, but using Julia 1.7 and all packages are available because of Pluto's built-in package manager

**Reply by gganapat** [reply]
The wasm backend on iPhone seems to be much faster than installed version on my Mac. Why is that? Plotting seems to be snappier also.

**Reply by VarLad** [reply]
@gganapat I'm pretty sure thats because the above link is using HTML + Plotly directly instead of Julia for Plotting

**Reply by fonsp** [reply]
Right! And Julia feels faster because it is interpreted, not JIT compiled.

**Comment by pkotlan** [comment]
hi everyone. it's been a few years since the last update here, i was wondering what the current state of wasm support for pluto is? have there been any major breakthroughs, or is there a newer discussion/issue where this is being tracked?

**Reply by frankier** [reply]
The newest effort is this: https://github.com/GroupTherapyOrg/Snapshot.jl
